internal/auth/internal/types/workload_identity.go

```diff
@@ -4,15 +4,50 @@
 package types
 
 import (
+	"github.com/hashicorp/consul/acl"
 	"github.com/hashicorp/consul/internal/resource"
 	pbauth "github.com/hashicorp/consul/proto-public/pbauth/v2beta1"
+	"github.com/hashicorp/consul/proto-public/pbresource"
 )
 
 func RegisterWorkloadIdentity(r resource.Registry) {
 	r.Register(resource.Registration{
-		Type:     pbauth.WorkloadIdentityType,
-		Proto:    &pbauth.WorkloadIdentity{},
-		Scope:    resource.ScopeNamespace,
+		Type:  pbauth.WorkloadIdentityType,
+		Proto: &pbauth.WorkloadIdentity{},
+		Scope: resource.ScopeNamespace,
+		ACLs: &resource.ACLHooks{
+			Read:  aclReadHookWorkloadIdentity,
+			Write: aclWriteHookWorkloadIdentity,
+			List:  aclListHookWorkloadIdentity,
+		},
 		Validate: nil,
 	})
 }
+
+func aclReadHookWorkloadIdentity(
+	authorizer acl.Authorizer,
+	authzCtx *acl.AuthorizerContext,
+	id *pbresource.ID,
+	res *pbresource.Resource,
+) error {
+	if id != nil {
+		return authorizer.ToAllowAuthorizer().IdentityReadAllowed(id.Name, authzCtx)
+	}
+	if res != nil {
+		return authorizer.ToAllowAuthorizer().IdentityReadAllowed(res.Id.Name, authzCtx)
+	}
+	return resource.ErrNeedData
+}
+
+func aclWriteHookWorkloadIdentity(
+	authorizer acl.Authorizer,
+	authzCtx *acl.AuthorizerContext,
+	res *pbresource.Resource) error {
+	return authorizer.ToAllowAuthorizer().IdentityWriteAllowed(res.Id.Name, authzCtx)
```

Comment:
I think nil writes are explicitly allowed according to https://github.com/hashicorp/consul/pull/18819

Reply:
if `res` is nil then wouldn't this `res.Id.Name`  throw a: `invalid memory address or nil pointer dereference`? I think that's what @cthain might be pointing out 🤔

Reply:
Oh 🤦 right, good catch!

Reply:
I'm actually not sure if `*pbresource.Resource` can ever be nil. None of the other hooks that I've searched have a nil check.

Reply:
`res` should not be nil because resource service validates it prior to calling ACL (and other hooks) https://github.com/hashicorp/consul/blob/41e6f6cd8be293fcd382ae8226f53c1428bfec53/agent/grpc-external/services/resource/write.go#L40

I don't think it'd would hurt to be defensive though.